<a href="https://colab.research.google.com/github/kahuroA/SQL-PRACTICE/blob/main/SQL_MINI_PROJECT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**SQL MINI-PROJECT** *

In [1]:
#load the sql extension
%load_ext sql
#create a temporary connection of sqlite
%sql sqlite://

'Connected: @None'

In [2]:
#import packages and read pandas dataframe in an sql table
import pandas as pd
with open('GrandElectors_by_state.csv','r') as f:
  Grand_Electors=pd.read_csv(f, index_col=0, encoding='utf-8')
%sql DROP TABLE IF EXISTS Grand_Electors;
%sql PERSIST Grand_Electors;
%sql SELECT * FROM Grand_Electors LIMIT 5;

 * sqlite://
Done.
 * sqlite://
 * sqlite://
Done.


[('Alabama', 9),
 ('Alaska', 3),
 ('Arizona', 11),
 ('Arkansas', 6),
 ('California', 55)]

In [3]:
#open the population csv with pandas and store and read the dataframe using sql table
with open('Population_by_state.csv','r') as f:
  Population=pd.read_csv(f, index_col=0, encoding='utf-8')
%sql DROP TABLE IF EXISTS Population;
%sql PERSIST Population;
%sql SELECT * FROM Population LIMIT 5;

 * sqlite://
Done.
 * sqlite://
 * sqlite://
Done.


[('CALIFORNIA', 39144818),
 ('TEXAS', 27469114),
 ('FLORIDA', 20271272),
 ('NEW YORK', 19795791),
 ('ILLINOIS', 12859995)]

In [4]:
# capitalize the states in the Grand_Electors table so as to enable an inner join
%sql UPDATE Grand_Electors SET State=UPPER(State);

 * sqlite://
51 rows affected.


[]

In [5]:
%%sql
--Make the inner join between Grand_Electors and Population, store the result in a new table called Combined
CREATE TABLE Combined AS
SELECT * FROM Grand_Electors
INNER JOIN Population  on Population.State = Grand_Electors.State;

 * sqlite://
Done.


[]

In [7]:
%%sql
--view the combined table
SELECT * FROM Combined LIMIT 5;

 * sqlite://
Done.


[('ALABAMA', 9, 'ALABAMA', 4858979),
 ('ALASKA', 3, 'ALASKA', 738432),
 ('ARIZONA', 11, 'ARIZONA', 6828065),
 ('ARKANSAS', 6, 'ARKANSAS', 2978204),
 ('CALIFORNIA', 55, 'CALIFORNIA', 39144818)]

In [8]:
%%sql
--We will drop the redundant column State:1 by creating a table Combined_new with relevant columns from Combined.
CREATE TABLE Combined_new (
  State  TEXT PRIMARY KEY,
  GrandElectors  int,
  Population  int
);
--After creating the table with relevant columns from Combine table, we will insert columns from Combined table 
INSERT INTO Combined_new SELECT State, GrandElectors, Population FROM Combined;
--We will then drop the table Combined
DROP TABLE Combined;
--finally we will alter the new table's name and give it the name Combined
ALTER TABLE Combined_new RENAME TO Combined;

 * sqlite://
Done.
Done.
Done.
Done.


[]

In [10]:
%%sql
--view the table combined
SELECT * FROM Combined LIMIT 5;

 * sqlite://
Done.


[('ALABAMA', 9, 4858979),
 ('ALASKA', 3, 738432),
 ('ARIZONA', 11, 6828065),
 ('ARKANSAS', 6, 2978204),
 ('CALIFORNIA', 55, 39144818)]

2. Your boss wants you to change the name of the "District of Columbia" state to its short version "DC". Please do that.

In [11]:
%%sql
--We update the state column with a condition of where we need to do the update
UPDATE Combined
SET State='DC'
WHERE State='DISTRICT OF COLUMBIA'

 * sqlite://
Done.


[]

3. To compute the ratio between the number of grand electors and the population. Please create a new column with that ratio.

In [12]:
%%sql
--create a new table with the ratio column
CREATE TABLE Combined_new AS
SELECT *, CAST(Combined.GrandElectors AS FLOAT)/CAST(Combined.Population AS FLOAT) AS Ratio
FROM   Combined;
--drop the old table Combined
DROP TABLE Combined;
--Rename the new table with ratio column as Combined
ALTER TABLE Combined_new Rename to Combined;

 * sqlite://
Done.
Done.
Done.


[]

In [13]:
%%sql
--Let us view the table with the new ratio column
SELECT * FROM Combined LIMIT 5;

 * sqlite://
Done.


[('ALABAMA', 9, 4858979, 1.8522409749043987e-06),
 ('ALASKA', 3, 738432, 4.0626625065002604e-06),
 ('ARIZONA', 11, 6828065, 1.610998137832607e-06),
 ('ARKANSAS', 6, 2978204, 2.0146370094190997e-06),
 ('CALIFORNIA', 55, 39144818, 1.4050391037710278e-06)]

4. To order the states by decreasing ratio of Grand Electors per capita. That will make our priority list

In [17]:
%%sql
CREATE TABLE Sorted AS
SELECT * FROM Combined
ORDER BY Ratio DESC;

 * sqlite://
Done.


[]

In [18]:
%%sql
--let us view the new table with ratio sorted in decreasing order
SELECT * FROM Sorted LIMIT 5;

 * sqlite://
Done.


[('WYOMING', 3, 586107, 5.118519314732634e-06),
 ('VERMONT', 3, 626042, 4.792010759661493e-06),
 ('DC', 3, 672228, 4.462771559649405e-06),
 ('ALASKA', 3, 738432, 4.0626625065002604e-06),
 ('NORTH DAKOTA', 3, 756927, 3.963394092164766e-06)]

5. To compute the running total of Grand Electors in that sorted list

In [24]:
%%sql
CREATE TABLE Sorted_Cummulative AS
SELECT
    t.State,t.GrandElectors,t.Population,t.Ratio,
    (SELECT Sum(r.GrandElectors) FROM Sorted as r WHERE t.Ratio <= r.Ratio) AS Cummulative
FROM Sorted as t
Group By t.Ratio ORDER BY t.Ratio DESC;

 * sqlite://
Done.


[]

In [33]:
%%sql
--Let us view the first 5 columns of the new table with the cummulative sum.
SELECT * FROM Sorted_Cummulative;

 * sqlite://
Done.


[('WYOMING', 3, 586107, 5.118519314732634e-06, 3),
 ('VERMONT', 3, 626042, 4.792010759661493e-06, 6),
 ('DC', 3, 672228, 4.462771559649405e-06, 9),
 ('ALASKA', 3, 738432, 4.0626625065002604e-06, 12),
 ('NORTH DAKOTA', 3, 756927, 3.963394092164766e-06, 15),
 ('RHODE ISLAND', 4, 1056298, 3.7868101615263876e-06, 19),
 ('SOUTH DAKOTA', 3, 858469, 3.4945932817609025e-06, 22),
 ('DELAWARE', 3, 945934, 3.1714686225466047e-06, 25),
 ('MAINE', 4, 1329328, 3.009039153617467e-06, 29),
 ('NEW HAMPSHIRE', 4, 1330608, 3.0061445594795764e-06, 33),
 ('MONTANA', 3, 1032949, 2.904306020916812e-06, 36),
 ('HAWAII', 4, 1431603, 2.794070702562093e-06, 40),
 ('WEST VIRGINIA', 5, 1844128, 2.7113085425740512e-06, 45),
 ('NEBRASKA', 5, 1896190, 2.636866558730929e-06, 50),
 ('IDAHO', 4, 1654930, 2.417020659484087e-06, 54),
 ('NEW MEXICO', 5, 2085109, 2.3979561739937817e-06, 59),
 ('NEVADA', 6, 2890845, 2.0755177119492743e-06, 65),
 ('KANSAS', 6, 2911641, 2.0606936088618066e-06, 71),
 ('ARKANSAS', 6, 2978204, 2.0146370094190997e-06, 77),
 ('MISSISSIPPI', 6, 2992333, 2.00512442966742e-06, 83),
 ('UTAH', 6, 2995919, 2.0027243727216926e-06, 89),
 ('CONNECTICUT', 7, 3590886, 1.949379623858847e-06, 96),
 ('IOWA', 6, 3123899, 1.9206766928124116e-06, 102),
 ('ALABAMA', 9, 4858979, 1.8522409749043987e-06, 111),
 ('SOUTH CAROLINA', 9, 4896146, 1.8381804790951905e-06, 120),
 ('MINNESOTA', 10, 5489594, 1.8216283389992047e-06, 130),
 ('KENTUCKY', 8, 4425092, 1.8078720171241637e-06, 138),
 ('OKLAHOMA', 7, 3911338, 1.7896689061390246e-06, 145),
 ('OREGON', 7, 4028977, 1.7374137405103083e-06, 152),
 ('WISCONSIN', 10, 5771337, 1.7327007589402595e-06, 162),
 ('LOUISIANA', 8, 4670724, 1.712796560019389e-06, 170),
 ('WASHINGTON', 12, 7170351, 1.673558240035948e-06, 182),
 ('TENNESSEE', 11, 6600299, 1.6665911650366142e-06, 193),
 ('MARYLAND', 10, 6006401, 1.6648905059785385e-06, 203),
 ('INDIANA', 11, 6619680, 1.6617117443743504e-06, 214),
 ('COLORADO', 9, 5456574, 1.6493865931260165e-06, 223),
 ('MISSOURI', 10, 6083672, 1.6437441071773757e-06, 233),
 ('MASSACHUSETTS', 11, 6794422, 1.6189750945702225e-06, 244),
 ('MICHIGAN', 16, 9922576, 1.6124844999927438e-06, 260),
 ('ARIZONA', 11, 6828065, 1.610998137832607e-06, 271),
 ('GEORGIA', 16, 10214860, 1.566345500574653e-06, 287),
 ('NEW JERSEY', 14, 8958013, 1.562846582160575e-06, 301),
 ('PENNSYLVANIA', 20, 12802503, 1.562194517743913e-06, 321),
 ('ILLINOIS', 20, 12859995, 1.5552105580134362e-06, 341),
 ('VIRGINIA', 13, 8382993, 1.5507587803067473e-06, 354),
 ('OHIO', 18, 11613423, 1.5499306276883225e-06, 372),
 ('NORTH CAROLINA', 15, 10042802, 1.4936070630487388e-06, 387),
 ('NEW YORK', 29, 19795791, 1.4649578791774474e-06, 416),
 ('FLORIDA', 29, 20271272, 1.4305959685213637e-06, 445),
 ('CALIFORNIA', 55, 39144818, 1.4050391037710278e-06, 500),
 ('TEXAS', 38, 27469114, 1.3833718845099992e-06, 538)]

6. Independently, to compute the half of the total of Grand Electors overall (in the whole country)

In [29]:
%%sql
SELECT SUM(GrandElectors)/2 FROM Sorted_Cummulative;

 * sqlite://
Done.


[(269,)]

7. To filter our sorted list of states in order to keep only the (top) ones enabling us to reach the computed threshold. (the other states can be ignored). That is our target list.

In [30]:
%%sql
CREATE TABLE Target_List AS
SELECT * FROM Sorted_Cummulative WHERE Cummulative<=269;

 * sqlite://
Done.


[]

In [31]:
%%sql
--Let us view our target list
SELECT * FROM Target_List

 * sqlite://
Done.


[('WYOMING', 3, 586107, 5.118519314732634e-06, 3),
 ('VERMONT', 3, 626042, 4.792010759661493e-06, 6),
 ('DC', 3, 672228, 4.462771559649405e-06, 9),
 ('ALASKA', 3, 738432, 4.0626625065002604e-06, 12),
 ('NORTH DAKOTA', 3, 756927, 3.963394092164766e-06, 15),
 ('RHODE ISLAND', 4, 1056298, 3.7868101615263876e-06, 19),
 ('SOUTH DAKOTA', 3, 858469, 3.4945932817609025e-06, 22),
 ('DELAWARE', 3, 945934, 3.1714686225466047e-06, 25),
 ('MAINE', 4, 1329328, 3.009039153617467e-06, 29),
 ('NEW HAMPSHIRE', 4, 1330608, 3.0061445594795764e-06, 33),
 ('MONTANA', 3, 1032949, 2.904306020916812e-06, 36),
 ('HAWAII', 4, 1431603, 2.794070702562093e-06, 40),
 ('WEST VIRGINIA', 5, 1844128, 2.7113085425740512e-06, 45),
 ('NEBRASKA', 5, 1896190, 2.636866558730929e-06, 50),
 ('IDAHO', 4, 1654930, 2.417020659484087e-06, 54),
 ('NEW MEXICO', 5, 2085109, 2.3979561739937817e-06, 59),
 ('NEVADA', 6, 2890845, 2.0755177119492743e-06, 65),
 ('KANSAS', 6, 2911641, 2.0606936088618066e-06, 71),
 ('ARKANSAS', 6, 2978204, 2.0146370094190997e-06, 77),
 ('MISSISSIPPI', 6, 2992333, 2.00512442966742e-06, 83),
 ('UTAH', 6, 2995919, 2.0027243727216926e-06, 89),
 ('CONNECTICUT', 7, 3590886, 1.949379623858847e-06, 96),
 ('IOWA', 6, 3123899, 1.9206766928124116e-06, 102),
 ('ALABAMA', 9, 4858979, 1.8522409749043987e-06, 111),
 ('SOUTH CAROLINA', 9, 4896146, 1.8381804790951905e-06, 120),
 ('MINNESOTA', 10, 5489594, 1.8216283389992047e-06, 130),
 ('KENTUCKY', 8, 4425092, 1.8078720171241637e-06, 138),
 ('OKLAHOMA', 7, 3911338, 1.7896689061390246e-06, 145),
 ('OREGON', 7, 4028977, 1.7374137405103083e-06, 152),
 ('WISCONSIN', 10, 5771337, 1.7327007589402595e-06, 162),
 ('LOUISIANA', 8, 4670724, 1.712796560019389e-06, 170),
 ('WASHINGTON', 12, 7170351, 1.673558240035948e-06, 182),
 ('TENNESSEE', 11, 6600299, 1.6665911650366142e-06, 193),
 ('MARYLAND', 10, 6006401, 1.6648905059785385e-06, 203),
 ('INDIANA', 11, 6619680, 1.6617117443743504e-06, 214),
 ('COLORADO', 9, 5456574, 1.6493865931260165e-06, 223),
 ('MISSOURI', 10, 6083672, 1.6437441071773757e-06, 233),
 ('MASSACHUSETTS', 11, 6794422, 1.6189750945702225e-06, 244),
 ('MICHIGAN', 16, 9922576, 1.6124844999927438e-06, 260)]

Add the first state for which the running total is larger than the threshold.

In [35]:
%%sql
--Let us add the first state for which the running total is larger than the threshold
INSERT INTO Target_List(State, GrandElectors, Population, Ratio, Cummulative)
SELECT * FROM Sorted_Cummulative WHERE State='ARIZONA';

 * sqlite://
Done.


[]

In [36]:
%%sql
--Let us view the Target List with the additional row
SELECT * FROM Target_List

 * sqlite://
Done.


[('WYOMING', 3, 586107, 5.118519314732634e-06, 3),
 ('VERMONT', 3, 626042, 4.792010759661493e-06, 6),
 ('DC', 3, 672228, 4.462771559649405e-06, 9),
 ('ALASKA', 3, 738432, 4.0626625065002604e-06, 12),
 ('NORTH DAKOTA', 3, 756927, 3.963394092164766e-06, 15),
 ('RHODE ISLAND', 4, 1056298, 3.7868101615263876e-06, 19),
 ('SOUTH DAKOTA', 3, 858469, 3.4945932817609025e-06, 22),
 ('DELAWARE', 3, 945934, 3.1714686225466047e-06, 25),
 ('MAINE', 4, 1329328, 3.009039153617467e-06, 29),
 ('NEW HAMPSHIRE', 4, 1330608, 3.0061445594795764e-06, 33),
 ('MONTANA', 3, 1032949, 2.904306020916812e-06, 36),
 ('HAWAII', 4, 1431603, 2.794070702562093e-06, 40),
 ('WEST VIRGINIA', 5, 1844128, 2.7113085425740512e-06, 45),
 ('NEBRASKA', 5, 1896190, 2.636866558730929e-06, 50),
 ('IDAHO', 4, 1654930, 2.417020659484087e-06, 54),
 ('NEW MEXICO', 5, 2085109, 2.3979561739937817e-06, 59),
 ('NEVADA', 6, 2890845, 2.0755177119492743e-06, 65),
 ('KANSAS', 6, 2911641, 2.0606936088618066e-06, 71),
 ('ARKANSAS', 6, 2978204, 2.0146370094190997e-06, 77),
 ('MISSISSIPPI', 6, 2992333, 2.00512442966742e-06, 83),
 ('UTAH', 6, 2995919, 2.0027243727216926e-06, 89),
 ('CONNECTICUT', 7, 3590886, 1.949379623858847e-06, 96),
 ('IOWA', 6, 3123899, 1.9206766928124116e-06, 102),
 ('ALABAMA', 9, 4858979, 1.8522409749043987e-06, 111),
 ('SOUTH CAROLINA', 9, 4896146, 1.8381804790951905e-06, 120),
 ('MINNESOTA', 10, 5489594, 1.8216283389992047e-06, 130),
 ('KENTUCKY', 8, 4425092, 1.8078720171241637e-06, 138),
 ('OKLAHOMA', 7, 3911338, 1.7896689061390246e-06, 145),
 ('OREGON', 7, 4028977, 1.7374137405103083e-06, 152),
 ('WISCONSIN', 10, 5771337, 1.7327007589402595e-06, 162),
 ('LOUISIANA', 8, 4670724, 1.712796560019389e-06, 170),
 ('WASHINGTON', 12, 7170351, 1.673558240035948e-06, 182),
 ('TENNESSEE', 11, 6600299, 1.6665911650366142e-06, 193),
 ('MARYLAND', 10, 6006401, 1.6648905059785385e-06, 203),
 ('INDIANA', 11, 6619680, 1.6617117443743504e-06, 214),
 ('COLORADO', 9, 5456574, 1.6493865931260165e-06, 223),
 ('MISSOURI', 10, 6083672, 1.6437441071773757e-06, 233),
 ('MASSACHUSETTS', 11, 6794422, 1.6189750945702225e-06, 244),
 ('MICHIGAN', 16, 9922576, 1.6124844999927438e-06, 260),
 ('ARIZONA', 11, 6828065, 1.610998137832607e-06, 271)]

How many states do you end up within the target list?

In [37]:
%%sql
SELECT COUNT(State) FROM Target_List;

 * sqlite://
Done.


[(40,)]

Is it a small or a large number? 

In [39]:
(40/51)*100

78.43137254901961

The total number of states in our target list is 78.4% of the total states in the US. So this is a large number.

 Do you think it would be a good recommendation to target those states?